x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2862 @@
+
+The current UI of the **Rule Updates** page has an button with the label "Update all" which currently simply updates all rules to their `target` version:
+
+![](2024-02-23-14-09-16.png)
+
+We can replace this button for an **Update rules with no conflicts** button. Clicking this button would make a request to `/upgrade/_perform` with a payload that includes only rules that have no conflicts. That payload can be built out of the response of the `/upgrade/_review` endpoint, including only rules with no conflicts and specifying that all rules should be updated to their `MERGED` version, which represents the "clean" (conflictless) update.
+
+```ts
+// [PSEUDOCODE]
+const reviewUpgradeResponse: ReviewRuleUpgradeResponseBody = await fetch(
+  '/internal/detection_engine/prebuilt_rules/upgrade/_review',
+  { method: 'POST' }
+);
+
+// Filter out rules with conflicts, and create payload pointing all to 'MERGE' version
+const performUpgradePayload = {
+  mode: 'SPECIFIC_RULES',
+  rules: reviewUpgradeResponse.rules
+    .filter((rule) => !rule.diff.has_conflict)
+    .map((rule) => ({
+      rule_id: rule.rule_id,
+      revision: rule.revision,
+      version: rule.version,
+    })),
+  pick_version: 'MERGE',
+};
+
+
+const performUpgradeResponse: PerformRuleUpgradeResponseBody = await fetch(
+  '/internal/detection_engine/prebuilt_rules/upgrade/_perform',
+  { method: 'POST', body: performUpgradePayload }
+);
+```
+
+#### Upgrading rules with conflicts
+
+Rules whose diffing algorithm resulted in a `CONFLICT` need to be manually resolved or confirmed by the user via the UI before making the API request to upgrade the rule.
+
+> See [Three-Way-Diff Component ticket](https://github.com/elastic/kibana/issues/171520) that details UI for solving conflicts.
+
+> See [designs](https://www.figma.com/file/gLHm8LpTtSkAUQHrkG3RHU/%5B8.7%5D-%5BRules%5D-Rule-Immutability%2FCustomization?type=design&mode=design&t=LkauhLzUKUatF6cL-0#712870904).
+
+## Other changes to UX/UI
+
+
+
+
+
+
+---- end of RFC -----
+
+
+## Scenario for bulk accepting updates
+
+- At rules table level
+  Bulk accept all rule updates that have no conflicting fields at all → This would make the request to update all these rules with no conflicts at all.
+  Accept all rule updates for non-customized rules.
+
+- At rule (filed-per-field) level
+  Accept all fields that have no conflicting changes, but still requires the user to manually solve fields with conflicts. → Once all fields are solved, the user can make the request to update the rule.
+  Accept all fields that are not customized
+
+- User Config
+  Create a user config/setting that the user can turn on to accept updates automatically if there are no conflicts.
+  Or we should more concretely how we want this setting to behave
+
+## Other open questions
```

Comment:
Should this text be removed or you're planning to finalize it?

Reply:
Removed